keymanapp/keyman
# fix(common/core/web): Remove empty rows in OSK

Replaces #5689 in addressing the KeymanWeb portion of #5327

If a touch keyboard had empty rows (compiled with older versions of Keyman Developer), the OSK would generate exceptions.
This cleans up how `ActiveLayout.correctLayerEmptyRowBug()` removes the empty rows.

For testing, I checked in the afghan_turkmen keyboard as included on the issue. I also created a custom `empty_row` keyboard in Developer 13 and edited the js file so there's an empty row in each of the 3 layers (different row ID).

## User Testing

* **TEST_ANDROID** Verifies OSK handles empty rows in the embedded app
Use an Android 5.0 emulator so the polyfill code is exercised.
1. Load the PR build of Keyman for Android
2. Test several of the default sil_euro_latin keys and verify they function correctly
3. Download and install [afghan_turkmen.kmp](https://darcywong00.github.io/examples/afghan/afghan_turkmen.kmp) which has an empty top row defined on the default layer
4. After the keyboard finished installing and switches, verify the afghan_turkmen OSK is fine (no empty rows)
5. Verify the default keys work. Note: since the attached keyboard came from the community site, it has an issue where the shift button doesn't switch layers

* **TEST_WEB** Verifies OSK 
Use the CI web test page artifacts using the new page `testing/empty-row/` in mobile viewport
1. Load the test page
2. With the empty_row keyboard, verify the default layer on the OSK is fine and keys function correctly
3. Use the shift to switch to the shift layer
4. Verify the shift layer OSK is fine and keys function correctly
5. Use the numeric key to switch to the numeric layer
6. Verify the numeric layer OSK is fine and keys function correctly
7. Use the globe key to switch to the afghan_turkmen keyboard
8. Verify the default layer on the afghan_turkmen OSK is fine (no empty rows)
9. Verify the keys work (except for the shift button, see the mentioned limitation in TEST_ANDROID)
